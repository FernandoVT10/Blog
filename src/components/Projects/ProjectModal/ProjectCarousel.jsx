export default ({ project }) => {
    return (
        <div
        id="projectCarousel"
        className="carousel project-details__carousel slide"
        data-ride="carousel">
            <ol className="carousel-indicators">
                {project.images.map((_, index) => {
                    const indicatorClass = index ? "" : "active";

                    return (
                        <li
                        key={index}
                        data-target="#projectCarousel"
                        data-slide-to={index}
                        className={indicatorClass}></li>
                    );
                })}
            </ol>
            <div className="carousel-inner">
                {project.images.map((image, index) => {
                    const carouselItemClass = index ? "" : "active";

                    return (
                        <div
                        className={`carousel-item ${carouselItemClass}`}
                        key={index}>
                            <img
                            className="d-block w-100 project-details__carousel-image"
                            src={`/img/projects/${image}`}
                            alt={project.title} />
                        </div>
                    );
                })}
                
            </div>
            <a
            className="carousel-control-prev project-details__control-button"
            href="#projectCarousel"
            role="button"
            data-slide="prev">
                <span
                className="carousel-control-prev-icon"
                aria-hidden="true"></span>
                <span className="sr-only">Previous</span>
            </a>
            <a
            className="carousel-control-next  project-details__control-button"
            href="#projectCarousel"
            role="button"
            data-slide="next">
                <span
                className="carousel-control-next-icon"
                aria-hidden="true"></span>
                <span className="sr-only">Next</span>
            </a>
        </div>
    );
};