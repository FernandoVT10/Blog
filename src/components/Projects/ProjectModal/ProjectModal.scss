@use "../../../styles/variables";

.project-details {
    &__carousel-image {
        width: 100%;
        height: 250px;
        object-fit: cover;
    }

    &__description {
        margin-top: 20px;
    }

    &__skill {
        cursor: pointer;
        background: variables.$secondary;
        color: black;
        margin-right: 10px;
        transition: .3s;

        &:hover {
            background: white;
        }
    }

    &__control-button {
        background: rgba($color: #000000, $alpha: .6);
        top: calc(50% - 40px);
        margin: 0 30px;
        height: 75px;
        width: 40px;
        opacity: 1;
        transition: .3s;

        &:hover {
            background: black;
        }
    }
}