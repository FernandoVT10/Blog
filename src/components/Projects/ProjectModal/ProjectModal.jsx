import Modal from "../../Modal";
import Loader from "../../Loader/";
import ProjectCarousel from "./ProjectCarousel";
import Api from "../../../ApiController";

import { useState, useEffect } from "react";
import Link from "next/link";
import { useRouter } from "next/router";

import "./ProjectModal.scss";

export default () => {
    const [project, setProject] = useState({ title: "Loading..." });
    const [modalActive, setModalActive] = useState(false);
    const [loading, setLoading] = useState(true);
    const [projectId, setProjectId] = useState(null);

    const router = useRouter();

    useEffect(() => {
        if(router.query.project) {
            if(projectId === router.query.project) {
                setModalActive(true);
            } else {
                setLoading(true);
                setProjectId(router.query.project);

                Api.get(`projects/getProjectById/${router.query.project}`)
                .then(project => {
                    if(!project) {
                        return;
                    }

                    setProject(project);
                    setModalActive(true);
                    setLoading(false);
                });
            }
        } else {
            setModalActive(false);
        }
    }, [router.query]);

    const onCloseModal = () => {
        setModalActive(false);

        const query = new URLSearchParams(location.search);

        // when the modal is closed, we add the modal-open class to the body
        document.body.classList.add("modal-open");

        router.push({
            pathname: router.pathname,
            query: { skill: query.get("skill") }
        });
    };

    const getProject = () => {
        if(loading || !project.skills) {
            return (
                <Loader active={true} />
            );
        }

        return (
            <div className="col-12">
                <ProjectCarousel project={project} />

                <p className="project-details__description">
                    { project.description }
                </p>

                <div className="project-details__skills">
                    {project.skills.map(({ _id, name }) => {
                        return( 
                            <Link href={`?skill=${name}`} key={_id}>
                                <span
                                className="badge badge-pill project-details__skill">
                                    { name }
                                </span>
                            </Link>
                        );
                    })}
                </div>
            </div>
        );
    };
    
    return (
        <Modal
        title={`${project.title}`}
        active={modalActive}
        modalSize="modal-lg"
        onClose={onCloseModal}
        prefix="project">
            <div className="container-fluid p-0">
                <div className="row project-details">
                    { getProject() }
                </div>
            </div>
        </Modal>
    );
};