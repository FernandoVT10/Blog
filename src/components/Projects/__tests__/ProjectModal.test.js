import ProjectModal from "../ProjectModal/";
import { act, Simulate } from "react-dom/test-utils";
import { render } from "react-dom";

const PROJECT_MOCK = {
    _id: 123,
    title: "Test title",
    description: "This is a test description",
    skills: [
        {
            _id: 21,
            name: "React JS"
        },
        {
            _id: 22,
            name: "Node JS"
        }
    ],
    images: ["test-image-1.jpg", "test-image-2.jpg", "test-image-3.jpg"]
};

window.$ = () => ({ modal: jest.fn(), on: jest.fn() });

jest.mock("next/router", () => ({
    useRouter: jest.fn(() => ({ query: { project: "123wqsa" } })),
}));

let container;

describe("<ProjectModal/> component", () => {
    beforeEach(() => {
        fetchMock.doMock();

        fetchMock.mockOnce(JSON.stringify(
            PROJECT_MOCK
        ));

        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it("Check if renders correctly", async () => {
        await act(async () => {
            render(<ProjectModal/>, container);
        });

        expect(container).toMatchSnapshot();
    });
});