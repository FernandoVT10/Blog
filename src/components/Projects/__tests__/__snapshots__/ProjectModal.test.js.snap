// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectModal/> component Check if renders correctly 1`] = `
<div>
  <div
    class="modal fade"
    id="project-modal"
    role="dialog"
    style="z-index: 1060;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-lg"
      role="document"
    >
      <div
        class="modal-content modal__content"
      >
        <div
          class="modal-header modal__header"
        >
          <h5
            class="modal__title"
          >
            Test title
          </h5>
          <button
            aria-label="Close"
            class="close modal__button"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <div
          class="modal-body modal__body"
        >
          <div
            class="container-fluid p-0"
          >
            <div
              class="row project-details"
            >
              <div
                class="col-12"
              >
                <div
                  class="carousel project-details__carousel slide"
                  data-ride="carousel"
                  id="projectCarousel"
                >
                  <ol
                    class="carousel-indicators"
                  >
                    <li
                      class="active"
                      data-slide-to="0"
                      data-target="#projectCarousel"
                    />
                    <li
                      class=""
                      data-slide-to="1"
                      data-target="#projectCarousel"
                    />
                    <li
                      class=""
                      data-slide-to="2"
                      data-target="#projectCarousel"
                    />
                  </ol>
                  <div
                    class="carousel-inner"
                  >
                    <div
                      class="carousel-item active"
                    >
                      <img
                        alt="Test title"
                        class="d-block w-100 project-details__carousel-image"
                        src="/img/projects/test-image-1.jpg"
                      />
                    </div>
                    <div
                      class="carousel-item "
                    >
                      <img
                        alt="Test title"
                        class="d-block w-100 project-details__carousel-image"
                        src="/img/projects/test-image-2.jpg"
                      />
                    </div>
                    <div
                      class="carousel-item "
                    >
                      <img
                        alt="Test title"
                        class="d-block w-100 project-details__carousel-image"
                        src="/img/projects/test-image-3.jpg"
                      />
                    </div>
                  </div>
                  <a
                    class="carousel-control-prev project-details__control-button"
                    data-slide="prev"
                    href="#projectCarousel"
                    role="button"
                  >
                    <span
                      aria-hidden="true"
                      class="carousel-control-prev-icon"
                    />
                    <span
                      class="sr-only"
                    >
                      Previous
                    </span>
                  </a>
                  <a
                    class="carousel-control-next  project-details__control-button"
                    data-slide="next"
                    href="#projectCarousel"
                    role="button"
                  >
                    <span
                      aria-hidden="true"
                      class="carousel-control-next-icon"
                    />
                    <span
                      class="sr-only"
                    >
                      Next
                    </span>
                  </a>
                </div>
                <p
                  class="project-details__description"
                >
                  This is a test description
                </p>
                <div
                  class="project-details__skills"
                >
                  <span
                    class="badge badge-pill project-details__skill"
                  >
                    React JS
                  </span>
                  <span
                    class="badge badge-pill project-details__skill"
                  >
                    Node JS
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
